gnuradio-runtime/python/gnuradio/gr/bindings/block_gateway_python.cc

```diff
@@ -7,17 +7,200 @@
  *
  */
 
+#include <gnuradio/api.h>
+#include <gnuradio/block.h>
+#include <gnuradio/io_signature.h>
+#include <string>
+
 #include <pybind11/complex.h>
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 
 namespace py = pybind11;
 
-#include <gnuradio/block_gateway.h>
+
+/*!
+ * The gateway block which performs all the magic.
+ *
+ * The gateway provides access to all the gr::block routines.
+ */
+typedef enum {
+    GW_BLOCK_GENERAL = 0,
+    GW_BLOCK_SYNC,
+    GW_BLOCK_DECIM,
+    GW_BLOCK_INTERP
+} gw_block_t;
+
+class block_gateway : virtual public gr::block
+{
+private:
+    py::handle _py_handle;
+
+public:
+    block_gateway(const py::handle& py_handle,
+                  const std::string& name,
+                  gr::io_signature::sptr in_sig,
+                  gr::io_signature::sptr out_sig)
+        : block(name, in_sig, out_sig)
+    {
+        _py_handle = py_handle;
+    }
+
+    // gr::block_gateway::sptr
+    typedef std::shared_ptr<block_gateway> sptr;
+
+    /*!
+     * Make a new gateway block.
+     * \param py_handle the pybind11 object with callback
+     * \param name the name of the block (Ex: "Shirley")
+     * \param in_sig the input signature for this block
+     * \param out_sig the output signature for this block
+     * \return a new gateway block
+     */
+    static sptr make(const py::object& py_handle,
```

Comment:
as a general rule, I'd recommend against in-line method definitions. I'd rather see the class definition as interface, and the method definitions out of line